src/ninja.cc

```diff
@@ -909,6 +912,38 @@ void printCompdb(const char* const directory, const Edge* const edge,
   printf("\"\n  }");
 }
 
+bool GetAllDependentEdges(Node* node, std::vector<Edge*>* depend_edges,
+                          std::set<Node*>* visited_nodes=NULL,
+                          std::set<Edge*>* visited_edges=NULL) {
+  if (visited_nodes == NULL || visited_edges == NULL) {
+    std::set<Node*> visited_nodes_;
+    std::set<Edge*> visited_edges_;
+    return GetAllDependentEdges(node, depend_edges, &visited_nodes_, &visited_edges_);
+  }
```

Comment:
It looks like this function is called in a loop with default arguments, each call will start with empty visited sets, so edges that are common to multiple calls will end up being appended to |depend_edges| multiple times. Is this what you really want here or a bug?

I suggest creating a small dedicated class to hold the visited state as an implementation detail instead, and make lifetime analysis clearer. E.g.:

```
struct DependentEdgeCollector {
  void CollectFrom(Node* node, std::vector<Edge*>*  out_edges) {
    ...
  }
  
  // Reset visited sets (only if this is needed, omit it otherwise).
  void Reset() {
    visited_nodes_.clear();
    visited_edges_.clear();
  }
 private:
  std::set<Node*> visited_nodes_;
  std::set<Edge*> visited_edges_;
};
```